torch/_tensor_str.py

```diff
@@ -65,132 +65,128 @@ def set_printoptions(
-            fmt_fn = format.format
-            sz = max(min_sz, 7 + prec)
-        else:
-            sz = max(min_sz, exp_max + 1 + include_decimal_int_mode)
-            format = '{:' + str(sz) + '.0f}'
-            fmt_fn = format.format
-            if include_decimal_int_mode:
-                format = '{:' + str(sz - 1) + '.0f}'
-                nonfinite_format = format + '.'
-                fmt_fn = _get_format_fn(format, nonfinite_format)
-    else:
-        if exp_max - exp_min > prec:
-            sz = 7 + prec
-            if abs(exp_max) > 99 or abs(exp_min) > 99:
-                sz = sz + 1
-            sz = max(min_sz, sz)
-            format = '{{:{}.{}e}}'.format(sz, prec)
-            fmt_fn = format.format
-        else:
-            if exp_max > prec + 1 or exp_max < 0:
-                sz = max(min_sz, 7)
-                scale = math.pow(10, max(exp_max - 1, _min_log_scale))
-            else:
-                if exp_max == 0:
-                    sz = 7
+                if exp_max - exp_min > PRINT_OPTS.precision or exp_max > 8 or exp_min < -4:
+                    value_str = ('{:' + str(self.max_width) +
+                                 '.' + str(PRINT_OPTS.precision) + 'e}').format(value)
+                    self.max_width = max(self.max_width, len(value_str))
+                    self.sci_mode = True
                 else:
-                    sz = exp_max + 6
-                sz = max(min_sz, sz)
-            format = '{{:{}.{}f}}'.format(sz, prec)
-            fmt_fn = format.format
-    return fmt_fn, scale, sz
+                    for value in tensor.tolist():
+                        value_str = ('{:.' +
+                                     str(PRINT_OPTS.precision) +
+                                     'f}').format(value)
+                        self.max_width = max(self.max_width, len(value_str))
+
+    def width(self):
+        return self.max_width
+
+    def format(self, value):
+        if self.floating_dtype:
+            if self.int_mode:
+                ret = '{:.0f}'.format(value)
+                if not (math.isinf(value) or math.isnan(value)):
+                    ret += '.'
+            elif self.sci_mode:
+                ret = ('{:' + str(self.max_width) +
+                       '.' + str(PRINT_OPTS.precision) + 'e}').format(value)
+            else:
+                ret = ('{:.' + str(PRINT_OPTS.precision) + 'f}').format(value)
+        else:
+            ret = '{}'.format(value)
+        return (self.max_width - len(ret)) * ' ' + ret
 
 
-def _scalar_str(self, fmt, scale):
-    scalar_str = fmt(self.item() / scale)
-    # The leading space for positives is ugly on scalars, so we strip it
-    return scalar_str.lstrip()
+def _scalar_str(self, formatter):
+    return formatter.format(self.item())
 
 
-def _vector_str(self, indent, fmt, scale, sz, summarize):
-    element_length = sz + 3
+def _vector_str(self, indent, formatter, summarize):
+    element_length = formatter.width() + 2
```

Comment:
can you comment what the magic number here is supposed to be for (I realize it was there before, but I don't remember what is was for before and I don't know what it's for now).